File: 2019/qualification_round/cryptopangrams.go
// Test set 1: Pass
// Test set 2: RE
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

func factor(n int, N int) (int, int) {
	if n%2 == 0 {
		return 2, n / 2
	}

	min := 3

	for i := min; i < N; i += 2 {
		if n%i == 0 {
			return i, n / i
		}
	}

	panic("factor panic")
}

func startpos(nums []int) int {
	for i, n := range nums {
		if n != nums[i+1] {
			return i
		}
	}

	panic("panic startpos")
}

func decrypt(nums []int) []string {
	copied := make([]int, len(nums))
	copy(copied, nums)
	sort.Sort(sort.IntSlice(copied))
	n2c := map[int]string{}
	count := 0
	for _, n := range copied {
		if _, exists := n2c[n]; !exists {
			n2c[n] = string('A' + count)
			count += 1
		}
	}
	result := []string{}
	for _, n := range nums {
		result = append(result, n2c[n])
	}
	return result
}

func solve(N int, L int, nums []int) string {
	result := make([]int, len(nums)+1)
	var index int

	start := startpos(nums)
	index = start

	a, b := factor(nums[start], N)
	nextn := nums[start+1]
	var now int
	if nextn%a == 0 {
		result[index] = b
		result[index+1] = a
		now = nextn / a
	} else {
		result[index] = a
		result[index+1] = b
		now = nextn / b
	}
	result[index+2] = now
	index += 3

	for _, n := range nums[start+2:] {
		now = n / now
		result[index] = now
		index += 1
	}

	now = result[start]
	rest := nums[:start]
	copiedRest := make([]int, len(rest))
	copy(copiedRest, rest)
	sort.Sort(sort.Reverse(sort.IntSlice(copiedRest)))
	index = start - 1
	for _, n := range copiedRest {
		now = n / now
		result[index] = now
		index -= 1
	}

	decrypted := decrypt(result)
	return strings.Join(decrypted, "")
}

func main() {
	stdin := bufio.NewScanner(os.Stdin)
	stdin.Scan()
	nTests, _ := strconv.Atoi(stdin.Text())
	for i := 0; i < nTests; i++ {
		stdin.Scan()
		nlTexts := strings.Fields(stdin.Text())
		N, _ := strconv.Atoi(nlTexts[0])
		L, _ := strconv.Atoi(nlTexts[1])
		nums := []int{}
		stdin.Scan()
		numTexts := strings.Fields(stdin.Text())
		for _, t := range numTexts {
			n, _ := strconv.Atoi(t)
			nums = append(nums, n)
		}
		ans := solve(N, L, nums)
		fmt.Printf("Case #%d: %s\n", i+1, ans)
	}
}

File: 2019/qualification_round/cryptopangrams.py
#!/usr/bin/env python

import string


def factor(n, N):
    if n % 2 == 0:
        return 2, n // 2

    min = 3
    # improve performance by this statements (but failed somewhy)
    # min = n // N
    # if min % 2 == 0:
    #     min += 1

    for i in range(min, N, 2):
        if n % i == 0:
            return i, n // i


def startpos(nums):
    for i, n in enumerate(nums):
        if n != nums[i + 1]:
            return i


def decrypt(ns):
    copied = sorted(list(set(ns)))
    alphas = string.ascii_uppercase
    n2c = {n: alphas[i] for i, n in enumerate(copied)}
    return [n2c[n] for n in ns]


def solve(N, L, nums):
    ns = []
    start = startpos(nums)

    a, b = factor(nums[start], N)
    nextn = nums[start + 1]
    if nextn % a == 0:
        ns.append(b)
        ns.append(a)
        now = nextn // a
        ns.append(now)
    else:
        ns.append(a)
        ns.append(b)
        now = nextn // b
        ns.append(now)

    for n in nums[start + 2:]:
        now = n // now
        ns.append(now)

    now = ns[0]
    for n in reversed(nums[0:start]):
        now = n // now
        ns.insert(0, now)

    return ''.join(decrypt(ns))


def solve_all():
    N_TESTS = int(input())
    for i in range(N_TESTS):
        N, L = map(int, input().split())
        nums = [int(x) for x in input().split()]
        ans = solve(N, L, nums)
        print("Case #%d: %s" % (i + 1, ans))


if __name__ == '__main__':
    solve_all()
